### ISSUE TITLE
Site API listing complaints (compared to old spring wiki)

### ISSUE THREAD

**Issue by sprunk** [beyond-all-reason/RecoilEngine#1430]
1) it's just a significantly worse layout. The page shows 3 callins in the space that the old page could show 10. Seeing nearby callins all at once helps you see related things which might be more like what you need, and it makes it much easier to scroll through and search for what you want. And the new one doesn't have the colored syntax highlighting.
![image](https://github.com/beyond-all-reason/spring/assets/2573076/8d54aa24-6cc6-4a03-bbb8-3cc309790609)
vs
![image](https://github.com/beyond-all-reason/spring/assets/2573076/34922070-3d68-4f1b-8dc9-10700e85349d)

2) the formatting for particularly complicated callins is broken/confusing
![image](https://github.com/beyond-all-reason/spring/assets/2573076/b43123d9-49b3-47cd-bf83-a0282273b3e9)
vs
![image](https://github.com/beyond-all-reason/spring/assets/2573076/c0727982-54a9-46bd-a2a9-5087b9906c1c)

**Comment by sprunk** [comment]
Def listings are also in a seemingly random order.

**Comment by badosu** [comment]
Navigation shouldn't rely on vertical compactness but on a usable navigation bar and proper sectioning.

Since we moved to emmylua json export and custom templating:

- we have full control over the template, I've improved the rendering of type definitions
- I've compacted the layout as much as I deemed reasonable.
- Nav is still unusable and there are no sections, but this is being tracked on https://github.com/beyond-all-reason/RecoilEngine/issues/1204 and https://github.com/beyond-all-reason/RecoilEngine/issues/1958 now.
- I'd rather new issues to be filed comparing to the current layout than working on this one.

Please reopen and let me know otherwise.

**Comment by sprunk** [comment]
> Please reopen and let me know otherwise.

cc @GoogleFrog (complaints were his, I just forwarded), listing is now at https://beyond-all-reason.github.io/RecoilEngine/docs/lua-api/ if you want to compare

**Comment by badosu** [comment]
Alright, I reopened and will track it on the aforementioned issue

**Comment by GoogleFrog** [comment]
What am I supposed to say? I clicked around for a few minutes and I'm not any happier. 

<img width="1089" height="520" alt="Image" src="https://github.com/user-attachments/assets/889435c2-6b99-4162-a4f5-e272ba3dcad1" />

All of these lines are redundant, they could just be written inline after `Spring.PlaySoundFile`. If there isn't space to write `number posx`, just then just write `posx`, because everyone knows that positions are numbers.

<img width="1356" height="161" alt="Image" src="https://github.com/user-attachments/assets/def2b3d5-440e-48fd-a545-7b7827b23e40" />

The Spring wiki found the space to put `number pox` inline, and even tells us what channel can be. I have to click away to find out:

<img width="626" height="59" alt="Image" src="https://github.com/user-attachments/assets/91b3068b-af96-4591-92b3-f71b0d8e978d" />

A glossary is nice, but it is a lot of clicking. And what does `0` and `...` mean?

The organisation is still terrible. The most relevant categories are {synced, unsynced}x{read, ctrl}. This all seems to appear under Spring. Putting everything on one page is a mistake because it makes the scrollbar near-unusable. Just look at this navigation pane, it is a half-page overview of what SyncedCtrl is, and I can click on it to go to each area. On one page.

<img width="408" height="517" alt="Image" src="https://github.com/user-attachments/assets/02cc2931-177c-4794-82dd-d0f37c1cabf1" />

**Comment by GoogleFrog** [comment]
<img width="1109" height="655" alt="Image" src="https://github.com/user-attachments/assets/7acc9250-b130-47b5-b74d-671af4935327" />

Spring wiki GetUnitLosState is still much more compact and useful. Having two identical headings confuses me, because when I find something, I assume everything is going to be there (see my previous comments about GetUnitLosState). This new version is probably more compact than earlier attempts, but _what is the bitmask_? The Spring wiki tells you. The links in the new wiki take you nowhere. The Spring wiki even found the space to enumerate all four non-raw table values, and to briefly explain each of them.

**Comment by badosu** [comment]
I mean... it's not like the current version is supposed to resolve all your concerns at this iteration, its a work in progress.

That said, I'd love to see a mockup of what is better. Would a layout that mimics spring wiki as identically as possible, _for type descriptions_ (remember sectioning and prose is still wip), be better for you?

After I finish certain features I'll make a layout identical to spring wiki. Personally I find it pretty messy and hard to read and reason about as a developer used to good documentation, but if that's what's deemed good for the target audience, I'll make it good.

**Comment by badosu** [comment]
> The Spring wiki tells you. The links in the new wiki take you nowhere. The Spring wiki even found the space to enumerate all four non-raw table values, and to briefly explain each of them.

I mean, it's not reasonable to pick certain instances one thing is better than the other in a crusade to try and prove something that is irrelevant to our goal (making documentation better), instead of just clarifying things that are good in one place that could help make the other better.

For example, I looked at spring wiki and I have absolutely no idea what these functions are supposed to do:

<img width="1712" height="1020" alt="Image" src="https://github.com/user-attachments/assets/5f66e614-1f65-4948-89ed-5a2e51d39113" />

Is this proof that spring wiki is terrible? No, just that it lacks good documentation for those functions, same as the other side. It's all a matter of putting an effort to make things better and providing _useful_ feedback.

I sent [a PR](https://github.com/beyond-all-reason/RecoilEngine/pull/2480) to adjust the documentation for those functions, thank you.

**Comment by badosu** [comment]
> What am I supposed to say? ...

Thank you, I had a plan to resolve the missing dependencies but I'll formalize them on the other issue.

- The `...` is our static analysis tool truncating long aliases, see [related issue resolved just a couple hours ago](https://github.com/EmmyLuaLs/emmylua-analyzer-rust/issues/623).
- The 0 is the same 0 you see on spring wiki

> The organisation is still terrible. The most relevant categories are {synced, unsynced}x{read, ctrl}. This all seems to appear under Spring. Putting everything on one page is a mistake because it makes the scrollbar near-unusable.

Literally two messages above yours **I** admit: "Nav is still unusable and there are no sections"

As _responded directly to you_ multiple times, over and over again, I am aware that organization is still terrible, I believe I also mentioned we're working to get there, it's already on the related epic. Currently blocked by [this issue on static analysis tool](https://github.com/EmmyLuaLs/emmylua-analyzer-rust/issues/623) 

You do realize it's very difficult to collaborate like this right? It's optimal to work under the assumption that the other read your message and minimally attempts to collaborate. Help me help you.

> I have to click away to find out:

I'll look at inlining aliases on the "spring wiki" layout

>  just then just write posx, because everyone knows that positions are numbers.

I hope this is a joke. Developers must have a crystal ball to understand an API on its documentation page based on what the parameter name _is meant to imply_ instead of _what it is_?

**Comment by GoogleFrog** [comment]
I told you I don't know what to say. I don't want to be involved in this, but I was ated so had a look and wrote something. We don't agree enough on what makes good documentation (or even usable documentation?) for me to collaborate.

> > just then just write posx, because everyone knows that positions are numbers.
>
> I hope this is a joke. Developers must have a crystal ball to understand an API on its documentation page based on what the parameter name is meant to imply instead of what it is?

Ok, perhaps this demonstrates something. Writing `number posx` is better than writing `posx`, but if you lack the room on the new layout for some reason, I'd be fine with you just writing `posx`. You've cut my sentence in half in a way which, if I were to be uncharitable, I would assume is meant to twist my meaning. I wrote:

> If there isn't space to write `number posx`, just then just write `posx`, because everyone knows that positions are numbers.

Look at the context. The line from the documentation below is full of useless `number` entries. If they were `posx` etc... then the line would be way more useful.

<img width="869" height="44" alt="Image" src="https://github.com/user-attachments/assets/f9f72914-256d-4bab-bc06-254ab80319e0" />

The names are listed below, but in a form that takes up a lot of space. This spends a lot of space for very little. Knowing that `posx` is a number tells developers barely anything. You complain about having to guess suggestively named variables, but _almost everything you want to know about `posx` is encoded in its suggestive name_. It is followed by `posy` and `posz`, this means that the three variables encode a coordinate. It is followed by three `speed{x,y,z}` variables, suggesting that the function takes a 3D position and a 3D speed vector. Neither version of the documentation says that `Spring.PlaySoundFile` requires a 3D position and speed, I had to read the variable names and use my brain. As far as I'm concerned, this is working as intended. Making the use of these numbers clearer would of course be fine, all else being equal, but there is very little extra clarity to be gained here, and everything added to the page has a cost.

But I really just don't want to be involved. Maybe the documentation will be usable when it leaves WIP. I don't know. Copying the Spring wiki verbatim would be... serviceable, but probably a dead end. But it is your project, do what you like with it.

**Comment by saurtron** [comment]
> Look at the context. The line from the documentation below is full of useless `number` entries. If they were `posx` etc... then the line would be way more useful.

I agree here the 'signature header' should either have just the variable names or both the type and variable name, not just the variable type.

Related to other concerns, I think discussing them on a half documented method maybe not the best. The long form of the parameters is supposed to have a description for each parameter. See for example [VBO:Upload](https://beyond-all-reason.github.io/RecoilEngine/docs/lua-api/#VBO:Upload).

Anyways I think maybe the expanded description about arguments and return could be in a collapsible section, so both a quick overview of all methods is possible as before, as well as more detailed information about them.

Other options would be having the details on mouseover when hovering over the parameters or return (maybe not too intuitive, and likely harder to implement), or having two pages as @badosu says, one with summaries and other with full details.

**Comment by badosu** [comment]
Its not _my_ project, it's ours one way or the other. The site has many major collaborators, the API has me and Rhys.

The fact there aren't more people interested in working on this doesn't mean I don't care about feedback, just feedback that is more usable.

We don't need spring wiki verbatim to have a spring wiki like template, if folks deem a layout better than another I'll pick what the community prefers, we want to service the game dev community, not mine or anyone's whims.

In any manner, even if the signal to noise ratio is low in feedbacks like this, the feedback was valuable to me, thank you

**Comment by badosu** [comment]
> > Look at the context. The line from the documentation below is full of useless `number` entries. If they were `posx` etc... then the line would be way more useful.
> 
> I agree here the 'signature header' should either have just the variable names or both the type and variable name, not just the variable type.
> 
> Related to other concerns, I think discussing them on a half documented method maybe not the best. The long form of the parameters is supposed to have a description for each parameter. See for example [VBO:Upload](https://beyond-all-reason.github.io/RecoilEngine/docs/lua-api/#VBO:Upload).
> 
> Anyways I think maybe the expanded description about arguments and return could be in a collapsible section, so both a quick overview of all methods is possible as before, as well as more detailed information about them.
> 
> Other options would be having the details on mouseover when hovering over the parameters or return (maybe not too intuitive, and likely harder to implement), or having two pages as @badosu says, one with summaries and other with full details.

Good feedback. I have an issue tracking improving the short signature block. Ill try and devise a compact layout documentation that goes in hand with the use case of "skimming through" docs

**Comment by badosu** [comment]
> You've cut my sentence in half in a way which, if I were to be uncharitable, I would assume is meant to twist my meaning.

Fair enough, you're right there. Not that it was meant to twist your meaning, but that it was an unfair selective response.

**Comment by rhys-vdw** [comment]
I agree that just the parameter name should be included in the function signature, as @GoogleFrog suggests. We ought to use a line break if it gets too long. The listing of parameters below can include type and param descriptions.

**Comment by rhys-vdw** [comment]
<img width="1864" height="1029" alt="Image" src="https://github.com/user-attachments/assets/7585ad44-da04-46b6-9ead-be198836d88f" />

Here I would suggest:
- list the parameter names only in the function at the top (this solves the problem of long union types, and gives a simple glance overview).
- always list the parameter name _before_ its type in the params list (very strange that this was not done)
- remove the blue link styling from the params--it's hurting legibility. In fact just remove the link entirely and only allow the function header itself to be linkable. Navigating to e.g. [this link](https://beyond-all-reason.github.io/RecoilEngine/docs/lua-api/#VBO:Upload-params.elemOffset) is bad because the context of the function is above the top of the scroll view. 

One thing that would vastly improve readability across the board is to remove the box from inline code elements:

<img width="736" height="421" alt="Image" src="https://github.com/user-attachments/assets/596cfe10-488b-4fe5-b6b1-3baa86d1d7c4" />

This could simply be a monospace font.

**Comment by rhys-vdw** [comment]
@GoogleFrog 

> Neither version of the documentation says that Spring.PlaySoundFile requires a 3D position and speed[...]

Please open a PR to add the missing description, you can find the source links in the docs to see where the doc comments are defined. Simple doc fixes like this are much better fixed when found rather than reported to others (if you know what they should be).  Otherwise open a separate issue for the particular issue with the documentation content.

**Comment by GoogleFrog** [comment]
> [...] I had to read the variable names and use my brain. As far as I'm concerned, this is working as intended.

**Comment by badosu** [comment]
Let me finish [this PR](https://github.com/beyond-all-reason/RecoilEngine/pull/2474) so I can bring in some improvements from updated emmylua doc cli fixes.

I want to improve the layout of the rendering, but my current priority should be to provide better sectioning, navigation and separation of API by environments. This should alleviate a lot of current issues before moving further to specific layout rendering.

That said, the templating for docgen is available at `doc/site/docgen` and can be issued via `mise run lua_pages`. I welcome PRs, I agree the code is very crude but I used the tools I could get the most done at the time.
